Persona: Scale presence to a max size when coinSize is specified

lynamemi opened the pull request:
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [X] Include a change request file using `$ npm run change`

#### Description of changes

- A recent [PR](https://github.com/OfficeDev/office-ui-fabric-react/pull/2912) introduced custom coin sizes. This PR scales the presence and presence icon proportionally to the coin size. 
- **_Update_**: the presence is now only scaled to a max size
- **_Update_**:  custom coinSize has been removed from examples to encourage users to use a predefined size

**Before:**
![image](https://user-images.githubusercontent.com/16619843/32511762-9b968ea2-c3a9-11e7-95ee-a3dda6b9344b.png)

**After:**
![2017-11-30-14-18-localhost_4322](https://user-images.githubusercontent.com/16619843/33459592-631a8dec-d5df-11e7-8631-fcc4b3e4ab10.png)
![2017-11-30-14-20-localhost_4322](https://user-images.githubusercontent.com/16619843/33459614-7974e0ce-d5df-11e7-9709-9e94cabfc609.png)


#### Focus areas to test

(optional)

lynamemi commented:
@MirandaGe can you take a look at this and let me know how it impacts your intended use case from your original PR? https://github.com/OfficeDev/office-ui-fabric-react/pull/2912

@mtennoe mentioned this change might be minor version bump worthy, but I am on the fence about it as it seems like a visual bug fix to me. Let me know your thoughts.

mikewheaton commented on packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.tsx:
```diff
@@ -17,15 +17,24 @@ export class PersonaPresence extends React.Component<IPersonaProps, {}> {
   public render(): JSX.Element | null {
-    let { presence } = this.props;
+    let { presence, coinSize } = this.props;
+    let coinSizeWithPresenceIconStyle = coinSize ? { fontSize: coinSize / 6, lineHeight: coinSize / 3 + 'px' } : undefined;
```
Why is `px` only needed for line height and not the others?

lynamemi replied:
Confirmed what we talked about offline that the others are defaulting to add the 'px' while lineHeight is not.

mtennoe replied:
How does that behave in various browsers? 
Adding px explicitly makes it clearer what the intention is

lynamemi commented:
Note that the suitcase icon disappears when I build from the main project directory but it comes back on the website build.

mtennoe commented on packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.tsx:
```diff
@@ -17,15 +17,24 @@ export class PersonaPresence extends React.Component<IPersonaProps, {}> {
 
   public render(): JSX.Element | null {
-    let { presence } = this.props;
+    let { presence, coinSize } = this.props;
```
These can all be const instead of let, since they dont change

mtennoe commented on packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.tsx:
```diff
@@ -17,15 +17,24 @@ export class PersonaPresence extends React.Component<IPersonaProps, {}> {
   public render(): JSX.Element | null {
-    let { presence } = this.props;
+    let { presence, coinSize } = this.props;
+    let coinSizeWithPresenceIconStyle = coinSize ? { fontSize: coinSize / 6, lineHeight: coinSize / 3 + 'px' } : undefined;
```
3 and 6 are magic numbers. Can you put them in consts with descriptive names to improve readability?

mtennoe commented on packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.tsx:
```diff
@@ -18,8 +18,10 @@ export class PersonaPresence extends React.Component<IPersonaProps, {}> {
     let { presence, coinSize } = this.props;
-    let coinSizeWithPresenceIconStyle = coinSize ? { fontSize: coinSize / 6, lineHeight: coinSize / 3 + 'px' } : undefined;
-    let coinSizeWithPresenceStyle = coinSize ? { width: coinSize / 3, height: coinSize / 3 } : undefined;
+    const coinSizeFontScaleFactor = 6;
```
Can move these two consts outside of the component class (underneath imports) to prevent the assignment from being done each time it calls render

mtennoe commented on packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.tsx:
```diff
@@ -17,15 +17,26 @@ export class PersonaPresence extends React.Component<IPersonaProps, {}> {
 
   public render(): JSX.Element | null {
-    let { presence } = this.props;
+    let { presence, coinSize } = this.props;
```
Can also be const

mikewheaton commented:
I'm not sure what's up with Screener.

lynamemi commented:
Me neither. I pulled master down this morning, which sometimes triggers problems like this. I'll ping @betrue-final-final to see what's up. No rush though, still need to hear from @MirandaGe before closing this.

betrue-final-final commented:
@dzearing said it's firing off false positives, or there's a shift in the screenshots without having any real pixel changes.

mtennoe commented on packages/office-ui-fabric-react/src/components/Persona/Persona.tsx:
```diff
@@ -105,7 +105,8 @@ export class Persona extends BaseComponent<IPersonaProps, {}> {
             PERSONA_PRESENCE[presence as PersonaPresenceEnum],
-            showSecondaryText && styles.showSecondaryText
+            showSecondaryText && styles.showSecondaryText,
+            coinSize && styles.coinSizeNoPresence
```
This logic belongs down in the PersonaCoin. PersonaCoin should take in a prop named "hidePresence", which hides the presence icon if it is set to true (by not including it in DOM at all). That prop is set to true if there is a custom coinsize passed in to Persona.tsx. This ensures a cleaner separation between the components.

lynamemi replied:
I just pushed an iteration where the PersonaCoin only renders a Presence if no coinSize is specified. Good call on not including it in the DOM. It's a pretty simple change - but I can iterate adding a "hidePresence" prop if you think it would add more value.

lynamemi commented:
ping @mtennoe  - Are you ok with these changes?

dzearing commented on packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx:
```diff
@@ -149,7 +149,7 @@ export class PersonaCoin extends React.Component<IPersonaProps, IPersonaState> {
               onLoadingStateChange={ this._onPhotoLoadingStateChange }
             />
-            <PersonaPresence { ...this.props } />
+            { !coinSize && <PersonaPresence { ...this.props } /> }
```
I don't quite think this is predictable. A customer who has a 20px coin, and needs a 26px would suddenly have the presence disappear. That's weird right?

lynamemi replied:
Matt had some great points from a design perspective about why they want to keep this behavior. I asked him to post here so we can continue the conversation.

dzearing replied:
Matt?

betrue-final-final commented:
Would we want presence to be a property, and by default not display at the large size, but display at smaller sizes?
On the custom size end, I would recommend just adding a new size. Giving people the variable of custom sizes could result in layouts being completely off the grid...literally off the 4px grid, and I don't think that makes even the author's job any easier. Let's have a sensible ramp for these, and have a new size.

micahgodbolt commented:
I agree this appears an odd solution, rather than just letting those that don't want the presence indicator to simply turn it off. As soon as we add logic to automatically turn it off, we'll need to add a prop to turn it back on.

lynamemi commented:
There is a strong desire from a design perspective to deprecate the custom coinSize prop all together. I spoke with the designer who needed the bigger coinSize, and a new size150 would meet that team's needs. As @betrue-final-final mentioned above, the size ramp is important to respect. We need to open this dialogue back up so we can address this properly.

lynamemi commented:
Update for this PR: I reverted back to a previous iteration where we were scaling the presence along with the coinSize. The update is that we now have a max size to fall back on so the presence size won't get ridiculously large. 

I also removed the coinSize example from the website to encourage users to use the predefined sizes.
